### PR TITLE
(#146) Rework Makefile-rules to use auto-variables $@, $< more widely

### PR DESCRIPTION
**Pull request by gapisback**
This commit refactors existing Makefile-rules to replace hard-coded
file-names (e.g. "certifier_proofs.cc") to instead use Make-provided
automatic variables (such as `$@` and `$<`).

In a few places, use:
- `$(<D)` to name the directory of the 1st pre-requisite
- `$(@D)` to name the directory of the target.

With this change, much of the rules become more compact and follow
a pattern consistent with Make-suggested usage of variables.

Refactor few mak-files to define common_objs list; remove redundant definitions of CFLAGS and other unused symbols / make-rules.

-----

NOTE: To reviewer: This change-set is built on top of PR #148 so that I could use the `-j n` feature to speed up dev-cycles while editing and testing Makefile-diffs. Ignore the diffs from the 1st commit. I have marked-up imp diffs arising from this specific change that merit some review / attention.

### CODE REVIEW

File: src/certifier.mak [hunk 1]
```diff
@@ -83,61 +83,61 @@ $(CL)/certifier.a: $(dobj)
 	$(AR) rcs $(CL)/certifier.a $(dobj)
 
 $(S)/certifier.pb.cc $(I)/certifier.pb.h: $(CP)/certifier.proto
-	$(PROTO) --cpp_out=$(S) --proto_path $(CP) $<
+	$(PROTO) --cpp_out=$(S) --proto_path $(<D) $<
```
Comment (gapisback) [inline src/certifier.mak:86]: The variable `$(<D)` means: use the dir of the 1st pre-requisite which here is $(CP).

This way, we more tightly bind the call to `protoc` to use the directory where the `certifier.proto` file lives as the  value of argument `--proto_path`.

File: src/certifier.mak [hunk 2]
```diff
@@ -83,61 +83,61 @@ $(CL)/certifier.a: $(dobj)
...
 	mv $(S)/certifier.pb.h $(I)
 
 $(O)/certifier_tests.o: $(S)/certifier_tests.cc $(I)/certifier.pb.h $(I)/certifier.h $(S)/test_support.cc
-	@echo "compiling certifier_tests.cc"
-	$(CC) $(CFLAGS) -c -o $(O)/certifier_tests.o $(S)/certifier_tests.cc
+	@echo "compiling $<"
+	$(CC) $(CFLAGS) -o $@ -c $<
```
Comment (gapisback) [inline src/certifier.mak:91]: This is the consistent pattern by which all changes are now done.

Use 

1. `$@` to refer to the target; i.e. `$(O)/certifier_tests.o`
2. `$<` to refer to the 1st pre-requisite file; i.e. `$(S)/certifier_tests.cc`

With these changes, the before / after make rules are now identical.

All make calls further below will be mostly exactly like this, barring an odd-ball case like the make call for `certifier.pb.cc` which needs an extra `-Wno-array-bounds` flag.

**Comment by gapisback** [comment]
hi, @jlmucb, @yelvmw -- As proposed and suggested last week, here is a draft PR showing the changes to just two Make-files, to replace hard-coded file names by using Make-provided automatic variables.

John -- I kicked this off for an initial preview look as I found myself updating Makefiles to add rules to handle generation of Python bindings using `swig` tool. I figured it would be good to lay down this Make-refactoring first, to ease the addition of new rules for the Python binding generation work.

File: utilities/cert_utility.mak [hunk 3]
```diff
@@ -54,15 +54,17 @@ AR=ar
 #export LD_LIBRARY_PATH=/usr/local/lib
 LDFLAGS= -L $(LOCAL_LIB) -lprotobuf -lgtest -lgflags -lpthread -L/usr/local/opt/openssl@1.1/lib/ -lcrypto -lssl
 
-dobj=	$(O)/cert_utility.o $(O)/certifier.pb.o $(O)/support.o $(O)/certifier.o \
-$(O)/certifier_proofs.o $(O)/simulated_enclave.o $(O)/application_enclave.o
+dobj =	$(O)/cert_utility.o $(O)/certifier.pb.o $(O)/support.o \
+        $(O)/certifier.o $(O)/certifier_proofs.o $(O)/simulated_enclave.o \
+        $(O)/application_enclave.o
```
Comment (gapisback) [inline utilities/cert_utility.mak:59]: Minor (unrelated) edit to indent the list of objects for improve readability.

File: utilities/cert_utility.mak [hunk 4]
```diff
@@ -54,15 +54,17 @@ AR=ar
...
-
-mobj=	$(O)/measurement_init.o $(O)/certifier.pb.o $(O)/support.o $(O)/certifier.o \
-$(O)/certifier_proofs.o $(O)/simulated_enclave.o $(O)/application_enclave.o
+key_dobj = $(O)/key_utility.o $(O)/certifier.pb.o $(O)/support.o \
+           $(O)/certifier.o $(O)/certifier_proofs.o $(O)/simulated_enclave.o \
+           $(O)/application_enclave.o
 
+mobj = $(O)/measurement_init.o $(O)/certifier.pb.o $(O)/support.o \
+       $(O)/certifier.o $(O)/certifier_proofs.o $(O)/simulated_enclave.o \
+       $(O)/application_enclave.o
```
Comment (gapisback) [inline utilities/cert_utility.mak:67]: Same for above two blocks .... Minor (unrelated) edit to indent the list of objects for improve readability.

File: utilities/cert_utility.mak [hunk 5]
```diff
@@ -72,53 +74,53 @@ clean:
 	rm -rf $(EXE_DIR)/cert_utility.exe
 
 cert_utility.exe: $(dobj) 
-	@echo "linking executable files"
-	$(LINK) -o $(EXE_DIR)/cert_utility.exe $(dobj) $(LDFLAGS)
+	@echo "linking executable $@"
```
Comment (gapisback) [inline utilities/cert_utility.mak:77]: Report the name of the executable being linked ...

File: application_service/app_service.mak [hunk 6]
```diff
@@ -33,7 +33,6 @@ I= $(SRC_DIR)/include
 INCLUDE= -I$(I) -I$(LIBSRC)/sev-snp -I/usr/local/opt/openssl@1.1/include/
 
 CFLAGS=$(INCLUDE) -O3 -g -Wall -std=c++11 -Wno-unused-variable -D X64 -D DEBUG -Wno-deprecated-declarations
-CFLAGS1=$(INCLUDE) -O1 -g -Wall -std=c++11 -Wno-unused-variable -D X64 -D DEBUG -Wno-deprecated-declarations
```
Comment (gapisback) [inline application_service/app_service.mak:36]: You will see some lines like this one being deleted. I found that definitions such as `CFLAGS1` are unused, so took the opportunity to cleanup the Makefile by deleting such unused definitions and unused targets (one case).

File: application_service/app_service.mak [hunk 7]
```diff
@@ -44,13 +43,18 @@ AR=ar
...
-$(O)/simulated_enclave.o $(O)/application_enclave.o $(O)/cc_helpers.o $(O)/cc_useful.o \
-$(O)/sev_support.o $(O)/sev_report.o
+dobj = $(O)/app_service.o $(O)/certifier.pb.o $(O)/certifier.o \
+       $(O)/certifier_proofs.o $(O)/support.o $(O)/simulated_enclave.o \
+       $(O)/application_enclave.o $(O)/cc_helpers.o $(O)/cc_useful.o \
+       $(O)/sev_support.o $(O)/sev_report.o
 
-user_dobj= $(O)/test_user.o $(O)/certifier.pb.o $(O)/certifier.o $(O)/certifier_proofs.o $(O)/support.o \
-$(O)/simulated_enclave.o $(O)/application_enclave.o $(O)/cc_helpers.o $(O)/cc_useful.o
+user_dobj = $(O)/test_user.o $(O)/certifier.pb.o $(O)/certifier.o \
```
Comment (gapisback) [inline application_service/app_service.mak:51]: Rearranged long-long lists of things to fit 80-char line limit which improves readability overall.

File: application_service/app_service.mak [hunk 8]
```diff
@@ -62,82 +66,82 @@ clean:
 	@echo "removing executable file"
 	rm -rf $(EXE_DIR)/app_service.exe
 
-hello_world.exe: hello_world.cc
-	@echo "hello_world.cc"
-	$(CC) $(CFLAGS) -o $(O)/hello_world.exe $(S)/hello_world.cc
+hello_world.exe: $(S)/hello_world.cc
+	@echo "compiling $<"
+	$(CC) $(CFLAGS) -o $(@D)/$@ -c $<
```
Comment (gapisback) [inline application_service/app_service.mak:71]: This is the standard pattern of change you will see ... on L70 and on L71. Just follow this pattern, to ease the review cycle.

File: application_service/app_service.mak [hunk 9]
```diff
@@ -62,82 +66,82 @@ clean:
...
 
-send_request.exe: $(O)/send_request.o $(O)/certifier.pb.o $(O)/certifier.o $(O)/support.o $(O)/application_enclave.o $(O)/simulated_enclave.o $(O)/certifier_proofs.o
-	@echo "send_request.exe"
-	$(LINK) -o $(EXE_DIR)/send_request.exe $(O)/send_request.o $(O)/certifier.pb.o \
-	$(O)/certifier.o $(O)/certifier_proofs.o $(O)/support.o $(O)/application_enclave.o $(O)/simulated_enclave.o $(LDFLAGS)
+$(EXE_DIR)/send_request.exe: $(send_req_dobj)
+	@echo "linking executable $@"
+	$(LINK) $(send_req_dobj) $(LDFLAGS) -o $(@D)/$@
```
Comment (gapisback) [inline application_service/app_service.mak:75]: Here's a 2nd instance of a change to consistently report the "Linking " message on L74 and the link command's arg sequence on L75.

File: application_service/app_service.mak [hunk 10]
```diff
@@ -62,82 +66,82 @@ clean:
...
 $(O)/send_request.o: $(S)/send_request.cc
-	@echo "send_request.cc"
-	$(CC) $(CFLAGS) -c -o $(O)/send_request.o $(S)/send_request.cc
+	@echo "\ncompiling $<"
+	$(CC) $(CFLAGS) -o $(@D)/$@ -c $<
 
-app_service.exe: $(dobj) 
-	@echo "linking executable files"
-	$(LINK) -o $(EXE_DIR)/app_service.exe $(dobj) $(LDFLAGS)
+$(EXE_DIR)/app_service.exe: $(dobj)
```
Comment (gapisback) [inline application_service/app_service.mak:81]: In some Makefiles, the target executable's output-dir was missing. Added it and adjusted the link command below.

File: application_service/app_service.mak [hunk 11]
```diff
@@ -62,82 +66,82 @@ clean:
...
+$(EXE_DIR)/test_user.exe: $(user_dobj)
+	@echo "linking executable $@"
+	$(LINK) $(user_dobj) $(LDFLAGS) -o $(@D)/$@
 
+$(I)/certifier.pb.h: $(US)/certifier.pb.cc
 $(US)/certifier.pb.cc: $(CP)/certifier.proto
-	$(PROTO) --proto_path=$(CP) --cpp_out=$(US) $<
-	mv certifier.pb.h $(I)
+	$(PROTO) --proto_path=$(<D) --cpp_out=$(@D) $<
+	mv $(@D)/certifier.pb.h $(I)
```
Comment (gapisback) [inline application_service/app_service.mak:92]: This is another instance of a pattern of change where protobuf generated source / output and compilation commands are rewritten to use symbols.

File: sample_apps/run_example.sh [hunk 12]
```diff
@@ -64,6 +64,14 @@ Local_lib_path=${LOCAL_LIB:-/usr/local/lib}
 # Sub-tools used in this script
 Jq=jq    # Needed for OE app; will be established later on.
 
+NumCPUs=1
+if [ "$(uname -s)" = "Linux" ]; then
+    NumCPUs=$(grep -c "^processor" /proc/cpuinfo)
+fi
+# Cap # of -j threads for make to 8
+NumMakeThreads=${NumCPUs}
+if [ "${NumMakeThreads}" -gt 8 ]; then NumMakeThreads=8; fi
```
Comment (gapisback) [inline sample_apps/run_example.sh:73]: [Unrelated diff from -j threads change set. Ignore this one.]

File: src/certifier.mak [hunk 13]
```diff
@@ -38,12 +38,12 @@ CL=..
...
+
+CFLAGS  = $(CFLAGS_COMMON) -O3
+
 ifdef ENABLE_SEV
-CFLAGS=$(INCLUDE) -O3 -g -Wall -std=c++11 -Wno-unused-variable -D X64 -D SEV_SNP -Wno-deprecated-declarations
-CFLAGS1=$(INCLUDE) -O1 -g -Wall -std=c++11 -Wno-unused-variable -D X64 -D SEV_SNP -Wno-deprecated-declarations
-else
-CFLAGS=$(INCLUDE) -O3 -g -Wall -std=c++11 -Wno-unused-variable -D X64 -Wno-deprecated-declarations
-CFLAGS1=$(INCLUDE) -O1 -g -Wall -std=c++11 -Wno-unused-variable -D X64 -Wno-deprecated-declarations
+CFLAGS  += -D SEV_SNP
```
Comment (gapisback) [inline src/certifier.mak:46]: Above block, simplifies `CFLAGS` definition by refactoring out common fragments and then conditionally building required `CFLAGS`.

File: utilities/policy_utilities.mak [hunk 14]
```diff
@@ -55,236 +53,237 @@ AR=ar
...
-$(O)/certifier.o $(O)/certifier_proofs.o $(O)/certifier.pb.o $(O)/simulated_enclave.o $(O)/application_enclave.o
-print_vse_clause_obj= $(O)/print_vse_clause.o $(O)/support.o $(O)/certifier.o $(O)/certifier_proofs.o \
-$(O)/certifier.pb.o $(O)/simulated_enclave.o $(O)/application_enclave.o
-print_signed_claim_obj= $(O)/print_signed_claim.o $(O)/support.o $(O)/certifier.o $(O)/certifier_proofs.o \
-$(O)/certifier.pb.o $(O)/simulated_enclave.o $(O)/application_enclave.o
-package_claims_obj= $(O)/package_claims.o $(O)/support.o $(O)/certifier.o $(O)/certifier_proofs.o \
-$(O)/certifier.pb.o $(O)/simulated_enclave.o $(O)/application_enclave.o
-print_packaged_claims_obj= $(O)/print_packaged_claims.o $(O)/support.o $(O)/certifier.o $(O)/certifier_proofs.o \
-$(O)/certifier.pb.o $(O)/simulated_enclave.o $(O)/application_enclave.o
+common_objs = $(O)/support.o $(O)/certifier.o $(O)/certifier_proofs.o \
```
Comment (gapisback) [inline utilities/policy_utilities.mak:56]: Refactored out common list of `.o`s into this symbol so the rest of the flow is easier to follow.

**Review by yelvmw — APPROVED** [review]
I focused on the specific patterns only instead of each of the repetitions. The change looks good to me. I believe our CI should be enough to catch any potential issues. Thanks!
